### PR TITLE
olexandr.vakulinskiy

### PR DESCRIPTION
**Pull request by AlexanderVak**

### CODE REVIEW

File: olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      @health += 1
+    elsif @hunger >= 7 || @thirst >= 7
+      @health -= 1
+    end
+    if @fatigue == 10
+      @health = 0
+    elsif  @fatigue >=7
+      @health -= 2
+    end
+
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: @olexandr.vakulinskiy there shouldn't be an empty line

File: olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  def drink
+    check_health
+    @thirst -= 1
+  end
+  def play
+    check_health
+    @fatigue += 2
+    @hunger += 1
+    @thirst += 1
+
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: remove empty line

File: olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  end
+  def get_characteristics
+    puts "PET CHARACTERISTICS
+Name: #{name}.
+Health: #{health} points.
+Hunger: #{hunger} points.
+Thirst: #{thirst} points.
+Fatigue: #{fatigue} points."
+  end
+
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: remove empty line

File: olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      @health -= 1
+    end
+    if @fatigue == 10
+      @health = 0
+    elsif  @fatigue >=7
+      @health -= 2
+    end
+
+  end
+  def feed
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: There should be a separation empty line between methods. Please, read more [here](https://github.com/rubocop-hq/ruby-style-guide#empty-lines-between-methods)

File: olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    check_health
+    @hunger -= 1
+  end
+  def drink
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: Should be a separation empty line

File: olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    check_health
+    @thirst -= 1
+  end
+  def play
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: Should be a separation empty line

File: olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    check_health
+    @fatigue += 2
+    @hunger += 1
+    @thirst += 1
+
+  end
+  def sleep
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: Should be a separation empty line

File: olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb [hunk 8]
```diff
@@ -0,0 +1,109 @@
+class Tamagotchi
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: file name must be lowercase, [separated by snake case if necessary](https://github.com/rubocop-hq/ruby-style-guide#snake-case-files)

File: olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    if @health <= 0
+      true
+    else
+      puts action_with_pet
+      false
+    end
+  end
+  def get_characteristics
+    puts "PET CHARACTERISTICS
+Name: #{name}.
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: please, use paddings for multi-line strings or [use heredoc](https://blog.saeloun.com/2020/04/08/heredoc-in-ruby-and-rails.html)
```suggestion
    puts "PET CHARACTERISTICS
             Name: #{name}.
             ...
```

Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/Tamagotchi.rb]: Much better for reading, if you'll use [heredoc](https://blog.saeloun.com/2020/04/08/heredoc-in-ruby-and-rails.html) here

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+  def is_dead?(action_with_pet)
+    if @health <= 0
+      true
+    else
+      puts action_with_pet
+      false
+    end
+  end
+  def get_characteristics
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: If you see a comment that is repeated in different places, you can make corrections that are described in the comment in places, where it can be missed. Please, add fixes here and below.

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    puts "PET CHARACTERISTICS
+Name: #{name}.
+Health: #{health} points.
+Hunger: #{hunger} points.
+Thirst: #{thirst} points.
+Fatigue: #{fatigue} points."
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: Much better for reading, if you'll use [heredoc ](https://blog.saeloun.com/2020/04/08/heredoc-in-ruby-and-rails.html)here

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    @fatigue = 0
+  end
+
+  def check_health
+    if @hunger == 0 || @thirst == 0
+      @health += 1
+    elsif @hunger >= 7 || @thirst >= 7
+      @health -= 1
+    end
+    if @fatigue == 10
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: please, use case here instead of two conditionals

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 13]
```diff
@@ -0,0 +1,111 @@
+class Tamagotchi
+
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: Good, you used empty lines after attribute accessor, but [she doesn't have to be between the name and accessor](https://github.com/rubocop-hq/ruby-style-guide#empty-lines-around-attribute-accessor)

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+  def initialize(name)
+    @name = name
+    @health = 10
+    @hunger = 2
+    @thirst = 2
+    @fatigue = 0
+  end
+
+  def check_health
+    if @hunger == 0 || @thirst == 0
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: better use [predicate methods](https://github.com/rubocop-hq/ruby-style-guide#predicate-methods) here and below
```suggestion
    if @hunger.zero? || @thirst.zero?
```

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+      @health += 1
+    elsif @hunger >= 7 || @thirst >= 7
+      @health -= 1
+    end
+    if @fatigue == 10
+      @health = 0
+    elsif  @fatigue >=7
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: there's extra whitespace here and no whitespace after `>=`

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+  end
+
+  def sleep
+    check_health
+    @fatigue -= 1
+    @hunger += 2
+    @thirst += 2
+  end
+
+  def is_dead?(action_with_pet)
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: use naming in [predicate methods without suffix](https://github.com/rubocop-hq/ruby-style-guide#predicate-methods-prefix)

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+Thirst: #{thirst} points.
+Fatigue: #{fatigue} points."
+  end
+end
+
+puts "Lets pick up a name for your new pet. Please enter a name"
+name = gets.chomp
+pet = Tamagotchi.new(name)
+
+def get_news(news_is_sad)
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: [Avoid prefixing method names with `get_`](https://github.com/rubocop-hq/ruby-style-guide#accessor_mutator_method_names)

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb (hunk 10 above)
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: [Avoid prefixing method names with `get_`](https://github.com/rubocop-hq/ruby-style-guide#accessor_mutator_method_names)

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+  news_is_sad
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: I think it's better not to return the argument here 🙂

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    puts "PET CHARACTERISTICS
+Name: #{name}.
+Health: #{health} points.
+Hunger: #{hunger} points.
+Thirst: #{thirst} points.
+Fatigue: #{fatigue} points."
+  end
+end
+
+puts "Lets pick up a name for your new pet. Please enter a name"
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: [use single quote](https://github.com/rubocop-hq/ruby-style-guide#single-quote) here and in similar places below

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+name = gets.chomp
+pet = Tamagotchi.new(name)
+
+def get_news(news_is_sad)
+  news_is_sad
+  if news_is_sad
+    puts "Sorry, your pet is dead =("
+  end
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: Prefer modifier `if/unless` usage when you have a [single-line body](https://github.com/rubocop-hq/ruby-style-guide#if-as-a-modifier)

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+You can play with your #{name} (enter play).
+You can send your #{name} to sleep (enter sleep).
+If you want to finish the program enter 'exit'"
+  puts '-' * 60
+  input = gets.chomp.downcase
+
+  case input
+  when "feed"
+    pet.feed
+    get_news(pet.is_dead?("You fed your pet.".chomp))
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: `.chomp` no need here

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+  if news_is_sad
+    puts "Sorry, your pet is dead =("
+  end
+end
+puts "Now you can have fun with your #{name}!"
+loop do
+  break if pet.is_dead?("")
+  pet.get_characteristics
+  puts '-' * 60
+  puts "You can feed your #{name} (enter 'feed').
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: Much better for reading, if you'll use heredoc here

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+  end
+
+  def sleep
+    check_health
+    @fatigue -= 1
+    @hunger += 2
+    @thirst += 2
+  end
+
+  def dead?(action_with_pet)
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: Note to you, predicate methods should only output boolean values (if you wish, you can correct the method and logic of the program now or not fixed it)

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+    pet.sleep
+    news(pet.dead?('Your pet is sleeping now'))
+  when 'exit'
+    break
+  else
+    puts 'That is not valid command!'
+  end
+end
+
+
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: one unnecessary whitespace here

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+    elsif @hunger >= 7 || @thirst >= 7
+      @health -= 1
+    end
+
+    case @fatigue
+    when 10
+      @health = 0
+    when 7...10
+      @health -= 2
+    end
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb:24]: I meant the other, you need to merge both conditions in one case 🙂 (if I understand correctly condition)

Comment (AlexanderVak) [reply olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb:24]: I'm not sure how to merge it, because in case must be different characteristic

Comment (hawkprovvv) [reply olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb:24]: @AlexanderVak You're right, I didn't notice that different attributes are checked 👍

File: olexandr.vakulinskiy/HW2/Task2/extended_string_class.rb [hunk 26]
```diff
@@ -0,0 +1,18 @@
+class String
+  def valid?
+    brackets = { "(" => ")", "<" => ">", "{" => "}", "[" => "]" }
+    brackets_stack = []
+    self.each_char do
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task2/extended_string_class.rb]: `self` [is not required here](https://github.com/rubocop-hq/ruby-style-guide#explicit-self)

File: olexandr.vakulinskiy/HW2/Task2/extended_string_class.rb [hunk 27]
```diff
@@ -0,0 +1,18 @@
...
+      |x| brackets_stack.push(x) if brackets.keys.include?(x)
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task2/extended_string_class.rb]: need fix
```suggestion
    each_char do |x|
      brackets_stack.push(x) if brackets.keys.include?(x)
```

Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task2/extended_string_class.rb]: more succinct use method `brackets.key?` instead `brackets.keys.include?(x)`

**Review by hawkprovvv — DISMISSED** [review]

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb (hunk 24 above)
Comment (hawkprovvv) [reply olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: @AlexanderVak please note

File: olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+
+  def initialize(max_num)
+    @num = max_num
+    @matrix = matrix
+    fill_matrix(primes)
+  end
+
+  private
+
+  def  primes
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb]: @AlexanderVak please note rubocop comments. There's unnecessary spacing.

File: olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+        @matrix = Matrix.rows(@matrix.to_a << row)
+      else
+        @matrix = Matrix.rows(@matrix.to_a << row.rotate(count))
+      end
+      count += 1
+    end
+  end
+end
+test_matrix = PrimeNumbers.new(10)
+p test_matrix.matrix
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb]: Please, add an [empty line](https://thoughtbot.com/blog/no-newline-at-end-of-file) at the end of all files.

File: olexandr.vakulinskiy/HW2/Task2/extended_string_class.rb [hunk 30]
```diff
@@ -0,0 +1,16 @@
+class String
+  def valid?
+    brackets = { '(' => ')', '<' => '>', '{' => '}', '[' => ']' }
+    brackets_stack = []
+    each_char do |x|
+      brackets_stack.push(x) if brackets.key?(x)
+      if brackets.values.include?(x)
+        brackets_stack.pop if brackets[brackets_stack.last] == x
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task2/extended_string_class.rb]: Here you can do everything in the one level condition through `&&`. [Nested conditions are bad.](https://github.com/rubocop-hq/ruby-style-guide#no-nested-conditionals)
```suggestion
      brackets_stack.pop if brackets.values.include?(x) && brackets[brackets_stack.last] == x
```

File: olexandr.vakulinskiy/HW2/Task2/extended_string_class.rb [hunk 31]
```diff
@@ -0,0 +1,15 @@
+class String
+  def valid?
+    brackets = { '(' => ')', '<' => '>', '{' => '}', '[' => ']' }
+    brackets_stack = []
+    each_char do |x|
+      brackets_stack.push(x) if brackets.key?(x)
+      brackets_stack.pop if brackets.values.include?(x) &&
+          brackets[brackets_stack.last] == x
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task2/extended_string_class.rb]: @AlexanderVak why the second part of the condition on the new line?

File: olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb]: @AlexanderVak according to the condition, the matrix output should be of the following form
```
2, 3, 5, 7
3, 5, 7, 2
5, 7, 2, 3
7, 2, 3, 5
```
Your output form:
`Matrix[[2, 3, 5, 7], [3, 5, 7, 2], [5, 7, 2, 3], [7, 2, 3, 5]]`

Please think about how you can change this. And pay attention to **ALL** unresolved comments 🙂

**Review by hawkprovvv — CHANGES_REQUESTED** [review]

File: olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb [hunk 33]
```diff
@@ -0,0 +1,113 @@
...
+  def sleep
+    check_health
+    @fatigue -= 1
+    @hunger += 2
+    @thirst += 2
+    puts 'Your pet is sleeping now.' unless @health == 0
+  end
+
+  def dead?
+    true ? @health <= 0 : false
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: This condition won't work. Change `true`

Comment (AlexanderVak) [reply olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: What do you mean?

Comment (hawkprovvv) [reply olexandr.vakulinskiy/HW2/Task1/tamagotchi.rb]: @AlexanderVak This condition `true ? @health <= 0 : false` will always work, because `true` - is always a positive condition that does not change. What is the point of such checking?

File: olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb [hunk 34]
```diff
@@ -0,0 +1,54 @@
+require 'matrix'
+class Matrix
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb]: U can, but I don't recommend adding methods to Ruby classes. In the future, it is better to create your own and work with it. You can [inherit your class](https://www.rubyguides.com/2019/01/what-is-inheritance-in-ruby/) from Matrix and do with it what u want 🙂

File: olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb [hunk 35]
```diff
@@ -0,0 +1,54 @@
...
+        print number.to_s + ', '
+        count += 1
+      else
+        print number.to_s + ' '
+        count += 1
+      end
+      if count == column_size
+        print "\n"
+        count = 0
+      end
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb]: @AlexanderVak Please, change these conditions to `case`. It will be much more readable.

File: olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb [hunk 36]
```diff
@@ -0,0 +1,54 @@
...
+      array.push(number) if (1..number).select { |d| (number % d).zero? } == [1, number]
+    end
+    array
+  end
+
+  def fill_matrix(row)
+    rotation_count = row.size - 1
+    count = 0
+    until count > rotation_count
+      @matrix = if count.zero?
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb:44]: Please, use [ternary operators](https://github.com/rubocop-hq/ruby-style-guide#ternary-operator) in such cases

File: olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb [hunk 37]
```diff
@@ -0,0 +1,54 @@
...
+                  Matrix.rows(@matrix.to_a << row)
+                else
+                  Matrix.rows(@matrix.to_a << row.rotate(count))
+                end
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb]: Please, use [ternary operators](https://github.com/rubocop-hq/ruby-style-guide#ternary-operator) in such cases.

File: olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb [hunk 38]
```diff
@@ -0,0 +1,54 @@
+require 'matrix'
+class Matrix
+  def print_matrix
+    count = 0
+    each do |number|
+      if count < column_size - 1
+        print number.to_s + ', '
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb]: Here and below in line 10 use [string interpolation](https://github.com/rubocop-hq/ruby-style-guide#string-interpolation).

File: olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb (hunk 35 above)
Comment (AlexanderVak) [reply olexandr.vakulinskiy/HW3/Task1/prime_numbers.rb]: How to set less than case in when?

File: olexandr.vakulinskiy/HW4/dash.rb [hunk 39]
```diff
@@ -0,0 +1,53 @@
...
+    when @num[-1]
+      condition_for_last(c)
+    else
+      condition_for_other(c)
+    end
+  end
+
+  def add_dashes_case
+    @num.each_char do |c|
+      case (c.to_i % 2).zero?
```
Comment (hawkprovvv) [inline olexandr.vakulinskiy/HW4/dash.rb]: @AlexanderVak use ternary operator here

**Review by hawkprovvv — APPROVED** [review]
